sdk/core/Azure.Core/src/Pipeline/HttpPipelineMessage.cs

```diff
@@ -60,5 +60,76 @@ public void SetProperty(string name, object value)
 
             _properties[name] = value;
         }
+
+        public Stream? PreserveResponseContent()
+        {
+            switch (_response?.ContentStream)
+            {
+                case ResponseShouldNotBeUsedStream responseContent:
+                    return responseContent.Original;
+                case Stream stream :
+                    _response.ContentStream = new ResponseShouldNotBeUsedStream(_response.ContentStream);
+                    return stream;
+                default:
+                    return null;
+            }
+        }
+
+        public void Dispose()
+        {
+            Request?.Dispose();
+            _response?.Dispose();
+        }
+
+        private class ResponseShouldNotBeUsedStream: Stream
```

Comment:
Would it be easier to modify Response.ContentStream's getter so you see the error as soon as you touch it instead of creating a Stream that you might hand off somewhere else for processing?

Reply:
The response is abstract so every transport would have to implement this. Not sure if we want to go in that direction.